src/relax.rs

```diff
@@ -0,0 +1,1084 @@
+                        v = (v << 4) | Self::unhex(it.next().unwrap());
+                        char::try_from(v)?
+                    }
+                    _ => return Err(Error::EscapeError(ch)),
+                };
+                s.push(decoded);
+            } else {
+                s.push(ch);
+            }
+        }
+        Ok(s)
+    }
+
+    fn handle_number(&self, pair: Pair<Rule>) -> Result<Document, Error> {
+        let text = pair.as_str();
+        let mut negative = false;
+        let t = if let Some(t) = text.strip_prefix('+') {
+            Self::syntax_error(!self.number_plus, "leading `+`", pair.as_span().start_pos())?;
+            t
+        } else if let Some(t) = text.strip_prefix('-') {
+            negative = true;
+            t
+        } else {
+            text
+        };
+        if t.starts_with("0x") || t.starts_with("0X") {
+            // Hexadecimal integer.
+            Self::syntax_error(
+                !self.number_hex,
+                "hexadecimal literal",
+                pair.as_span().start_pos(),
+            )?;
+            return Self::from_str_radix(&t, 16, negative);
+        } else if t.starts_with("0b") || t.starts_with("0B") {
+            // Hexadecimal integer.
+            Self::syntax_error(
+                !self.number_bin,
+                "binary literal",
+                pair.as_span().start_pos(),
+            )?;
+            return Self::from_str_radix(&t, 2, negative);
+        } else if t.starts_with("0o") || t.starts_with("0O") {
+            // Hexadecimal integer.
+            Self::syntax_error(
+                !self.number_oct,
+                "octal literal",
+                pair.as_span().start_pos(),
+            )?;
+            return Self::from_str_radix(&t, 8, negative);
+        } else if t.contains('.')
+            || t.contains('e')
+            || t.contains('E')
+            || t == "NaN"
+            || t == "Infinity"
+        {
+            // Floating point number.
+            Self::syntax_error(
+                !self.number_lax_dec_point && (t.starts_with('.') || t.ends_with('.')),
+                "bad float literal",
+                pair.as_span().start_pos(),
+            )?;
+            return Ok(Document::Float(text.parse().unwrap()));
+        } else {
+            // Decimal integer.
+            return Self::from_str_radix(t, 10, negative);
+            /*
+            let val = u128::from_str_radix(t, 10).unwrap();
+            if negative {
+                let val = -(val as i128);
+                return Ok(Document::Int(Int::new(val, Base::Dec)));
+            } else {
+                return Ok(Document::Int(Int::new(val, Base::Dec)));
+            }
+            */
+        }
+    }
+
+    fn handle_kvpair(&self, pairs: &mut Pairs<Rule>) -> Result<(Document, bool), Error> {
+        let mut k = usize::MAX;
+        let mut v = usize::MAX;
+        let mut kv = vec![];
+        let mut comma = false;
+        loop {
```

Comment:
Both of these loops are using `peek` to look at the next parsed rule to decide whether or not to consume the rule on the current iteration of the loop (e.g. current kvpair) or the next iteration of the loop.  This is done to heuristically guess to which element in the aggregate the comments belong.

Because of the peek behavior, the loop cannot be turned into a `for` loop, but it _can_ be simplified into a `while let` loop.